src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs

```diff
@@ -548,302 +424,118 @@ await foreach (var index in _source.GetIndexesAsync())
             }
-        }
-
-        private void AddInfoToSmugglerResult(SmugglerResult result, string message)
-        {
-            result.AddInfo(message);
-            _onProgress.Invoke(result.Progress);
-        }
-
-        private void SetDocumentOrTombstoneFlags(ref DocumentFlags flags, ref NonPersistentDocumentFlags nonPersistentFlags, BuildVersionType buildType)
-        {
-            flags = flags.Strip(DocumentFlags.FromClusterTransaction | DocumentFlags.FromReplication);
-            nonPersistentFlags |= NonPersistentDocumentFlags.FromSmuggler;
 
-            if (_options.SkipRevisionCreation)
-                nonPersistentFlags |= NonPersistentDocumentFlags.SkipRevisionCreationForSmuggler;
-
-            switch (buildType)
-            {
-                case BuildVersionType.V4:
-                case BuildVersionType.V5:
-                case BuildVersionType.GreaterThanCurrent:
-                    {
-                        if (_options.OperateOnTypes.HasFlag(DatabaseItemType.RevisionDocuments) == false)
-                            flags = flags.Strip(DocumentFlags.HasRevisions);
-
-                        // those flags will be re-added once counter/time-series is imported
-                        flags = flags.Strip(DocumentFlags.HasCounters);
-                        flags = flags.Strip(DocumentFlags.HasTimeSeries);
-
-                        // attachments are special because they are referenced
-                        if (_options.OperateOnTypes.HasFlag(DatabaseItemType.Attachments) == false)
-                            flags = flags.Strip(DocumentFlags.HasAttachments);
-
-                        break;
-                    }
-            }
+            return counts;
         }
 
-        private bool SkipDocument(BuildVersionType buildType, bool isPreV4Revision, DocumentItem item, SmugglerResult result, ref List<string> legacyIdsToDelete)
+        protected override async Task<SmugglerProgressBase.Counts> ProcessCompareExchangeAsync(SmugglerResult result)
         {
-            if (buildType == BuildVersionType.V3 == false)
-                return false;
-
-            if (_options.OperateOnTypes.HasFlag(DatabaseItemType.RevisionDocuments) == false && isPreV4Revision)
-            {
-                result.Documents.SkippedCount++;
-                if (result.Documents.SkippedCount % 1000 == 0)
-                    AddInfoToSmugglerResult(result, $"Skipped {result.Documents.SkippedCount:#,#;;0} documents.");
-
-                return true;
-            }
+            bool isSharded = false;
+            int index = -1;
+            DatabaseRecord shardedRecord = null;
+            result.CompareExchange.Start();
 
-            if ((item.Document.NonPersistentFlags & NonPersistentDocumentFlags.LegacyDeleteMarker) == NonPersistentDocumentFlags.LegacyDeleteMarker)
+            //Handle compare exchange for shard database export
+            if (_source is DatabaseSource)
```

Comment:
something is not right, we should avoid having code like that in the databaseSmuggler

Reply:
done